### PR TITLE
o/i/apparmorprompting: add prompting `common` package

### PR DESCRIPTION
**Pull request by olivercalder**
This PR has been split into three PRs, some of which have/will be further split into smaller PRs:
1. https://github.com/snapcore/snapd/pull/13849 (common types and helpers)
2. https://github.com/snapcore/snapd/pull/13850 (constraints and abstract permissions)
3. https://github.com/snapcore/snapd/pull/13730 (path patterns -- to be split)

--------

Original description:

Add the `common` package, which provides type definitions and functions which are used by other 
components related to apparmor prompting.

Notably, this package handles:
- Interface-specific constraints
- Abstracted permissions (again, interface-specific)
- ~~Path pattern matching~~ moved to #13730 
- ~~Path pattern validation~~ moved to #13730
- ~~Path pattern precedence~~ moved to #13730
- ~~Expansion of groups in path patterns~~ moved to #13730 

The meta-PR for apparmor prompting can be found here, if it would be helpful to see how this package is used: https://github.com/snapcore/snapd/pull/12964

The bulk of the work to support arbitrary wildcards and groups in path patterns was completed here: https://github.com/olivercalder/snapd/pull/30
There was a very helpful review by @TheSignPainter98 which I addressed as best I could, and I'm hoping any future comments about those components can be consolidated in this PR.

### CODE REVIEW

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+	currPriority    priorityType
+	nextPatternsMap map[string]int
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, pattern string, e int) {
+	if priority < np.currPriority {
+		return
+	}
+	if priority > np.currPriority {
+		np.nextPatternsMap = make(map[string]int)
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:516]:
```go
np.nextPatterns = np.nextPatterns[:0]
```

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 2]
```diff
@@ -0,0 +1,789 @@
...
+	priorityTerminated
+	priorityLiteral
+)
+
+type nextPatternsContainer struct {
+	currPriority    priorityType
+	nextPatternsMap map[string]int
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, pattern string, e int) {
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:511]:
```go
func (np *nextPatternsContainer) addWithPriority(priority priorityType, patternReader *strings.Reader) {

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 3]
```diff
@@ -0,0 +1,789 @@
...
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
+	if len(patterns) == 0 {
+		return "", ErrNoPatterns
+	}
+	// Map pattern to number of escaped characters which have been seen
+	remainingPatterns := make(map[string]int, len(patterns))
+	for _, pattern := range patterns {
+		remainingPatterns[pattern] = 0
+	}
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:582]: Make map of unique patterns here, and maybe check that each doesn't end in `'\\'`. Then, do the following:
```go
remainingPatterns := make([]*strings.Reader, len(uniquePatterns))
for unique := range uniquePatterns {
	remainingPatterns = append(remainingPatterns, &strings.NewReader(unique))
}
```

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+	// Loop over index into each pattern until only one pattern left
+	for i := 0; len(remainingPatterns) > 1; i++ {
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:584]: A simple `for len(remainingPatterns) > 1 {` should work here.

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+		nextPatterns := nextPatternsContainer{}
+		for pattern, e := range remainingPatterns {
+			// For each pattern, e is number of escaped chars, thus the number
+			// which should be added to the index to compare equivalent indices
+			// into all patterns
+			if i+e == len(pattern) {
+				nextPatterns.addWithPriority(priorityTerminated, pattern, e)
+				continue
+			}
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:593]:
```go
for _, patternReader := range remainingPatterns {
	r, _, err := patternReader.ReadRune()
	if err != nil {
		nextPatterns.AddWithPriority(priorityTerminated, patternReader)
		continue
	}
	...
}

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 6]
```diff
@@ -0,0 +1,789 @@
...
+			// Check for '?' and '*' before '\\', since "\\*" is literal '*', etc.
+			if pattern[i+e] == '?' {
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:595]: Here and below, simply check the previously-read rune `r`.

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 7]
```diff
@@ -0,0 +1,789 @@
...
+				}
+				nextPatterns.addWithPriority(priorityGlob, pattern, e)
+				continue
+			}
+			if pattern[i+e] == '\\' {
+				e += 1
+				if i+e == len(pattern) {
+					return "", fmt.Errorf(`invalid path pattern: trailing '\' character: %q`, pattern)
+				}
+			}
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:613]: Here, just read (and record) another rune:
```go
if r == '\\' {
	r, _, err = patternReader.ReadRune()
	if err != nil {
		return err...
	}
}

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 8]
```diff
@@ -0,0 +1,789 @@
...
+				continue
+			}
+			// pattern[i+e:i+e+2] must be "/*"
+			if i+e+2 >= len(pattern) || pattern[i+e+2] != '*' {
+				// pattern[i+e:i+e+3] must not be "/**"
+				nextPatterns.addWithPriority(prioritySinglestar, pattern, e)
+				continue
+			}
+			// pattern[i+e:i+e+3] must be "/**"
+			if i+e+3 == len(pattern) || (pattern[i+e+3] == '/' && (i+e+4 == len(pattern) || (i+e+5 == len(pattern) && pattern[i+e+4] == '*'))) {
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:627]: Rework these to be careful about reading runes to look ahead. Probably read into a new reader instead, and use `ReadRune()` on that buffer to advance.

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 9]
```diff
@@ -0,0 +1,789 @@
...
+func PathPatternMatch(pattern string, path string) (bool, error) {
+	// Check the usual doublestar match first, in case the pattern is malformed
+	// and causes an error, and return the error if so.
+	matched, err := doublestar.Match(pattern, path)
+	if err != nil {
+		return false, err
+	}
+	// No matter if doublestar matched, return false if pattern ends in '/' but
+	// path is not a directory.
+	if len(pattern) > 0 && pattern[len(pattern)-1] == '/' && len(path) > 0 && path[len(path)-1] != '/' {
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/common/common.go:780]: Use `strings.HasSuffix()` to make sure trailing runes are `'/'`, rather than just trailing bytes.

**Review by MiguelPires — COMMENTED** [review]
Some comments and questions. I didn't do a full pass on the pattern expansion. I started yesterday but, since you're rewriting it, I might as well review the new version. Also IMO this PR is a bit too big, the path pattern expansion could be its own PR for instance

**Comment by pedronis** [comment]
>  Also IMO this PR is a bit too big, the path pattern expansion could be its own PR for instance

if this PR could be split in two I agree it would likely be good

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 10]
```diff
@@ -0,0 +1,789 @@
...
+			if r == '*' {
+				return -1
+			}
+			return r
+		}
+		return strings.Map(deleteStars, s) + "*"
+	})
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-2]
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:487]: Yes, see: https://cs.opensource.google/go/go/+/refs/tags/go1.22.1:src/strings/strings.go;l=1295

```go
return s[:len(s)-len(suffix)]
```

where `suffix` here would be `"/*"`.

**Comment by olivercalder** [comment]
This PR has been split, and all path pattern expansion, matching, and precedence has been moved to PR #13730.

When #13730 is merged, various TODO items introduced in this PR should be addressed.

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 11]
```diff
@@ -0,0 +1,789 @@
...
+}
+
+// Returns a new unique ID.
+// The ID is the current unix time in nanoseconds encoded as base32.
+func NewID() string {
+	nowUnix := uint64(time.Now().UnixNano())
+	nowBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(nowBytes, nowUnix)
+	id := base32.StdEncoding.EncodeToString(nowBytes)
+	return id
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:161]: I did this to make it clear that the base32-encoded time is the ID, but would also be fine with your suggestion

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 12]
```diff
@@ -0,0 +1,789 @@
...
+var (
+	ErrInvalidSnapLabel           = errors.New("the given label cannot be converted to snap")
+	ErrNoPatterns                 = errors.New("no patterns given, cannot establish precedence")
+	ErrPermissionNotInList        = errors.New("permission not found in permissions list")
+	ErrPermissionsListEmpty       = errors.New("permissions list empty")
+	ErrUnrecognizedFilePermission = errors.New("file permissions mask contains unrecognized permission")
+)
+
+type Constraints struct {
+	PathPattern string   `json:"path-pattern"`
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go]: They are serialized when writing constraints to disk (as part of a rule, so it persists across restarts) or when sending rules over the request rules API endpoint. I agree, these should be omitempty.

The original reasoning was that there would be different constraint types for different mediation classes of requests (see https://github.com/snapcore/snapd/blob/master/sandbox/apparmor/notify/message.go) depending on the interface. But I think the constraints type will actually be shared, with the validation occurring on a per-interface basis. So these should definitely be omitempty.

**Comment by olivercalder** [comment]
Rebased on master to pull in various improvements.

**Comment by olivercalder** [comment]
This PR is independent of the common package path patterns PR #13730, and is ready for review @zyga @bboozzoo @MiguelPires

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+		c.Permissions = c.Permissions[:len(c.Permissions)-1]
+	}
+	if origLen == len(c.Permissions) {
+		return ErrPermissionNotInList
+	}
+	return nil
+}
+
+// ContainPermissions returns true if the constraints include every one of the
+// givne permissions.
```
Comment (MiguelPires) [inline overlord/ifacestate/apparmorprompting/common/common.go:96]:
```suggestion
// given permissions.
```

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+type OutcomeType string
+
+const (
+	OutcomeUnset OutcomeType = ""
```
Comment (MiguelPires) [inline overlord/ifacestate/apparmorprompting/common/common.go:109]: Nitpick: maybe a value similar to this would make error messages a bit clearer? Otherwise an error message could be `must be "allow" or "deny": ""` 
```suggestion
	OutcomeUnset OutcomeType = "(unset)"
```

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 15]
```diff
@@ -0,0 +1,466 @@
...
+		return false, nil
+	default:
+		return false, fmt.Errorf(`invalid outcome: must be %q or %q: %q`, OutcomeAllow, OutcomeDeny, outcome)
+	}
+}
+
+type LifespanType string
+
+const (
+	LifespanUnset    LifespanType = ""
```
Comment (MiguelPires) [inline overlord/ifacestate/apparmorprompting/common/common.go:129]: Same question I guess. Whatever you prefer for the previous one, we should keep this consistent

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+// interface, otherwise returns an error.
+func (c *Constraints) ValidateForInterface(iface string) error {
+	switch iface {
+	case "home", "camera":
+		// TODO: change to this once PR #13730 is merged:
+		// if err := ValidatePathPattern(c.PathPattern); err != nil {
+		//	return err
+		// }
+	default:
+		return fmt.Errorf("constraints incompatible with the given interface: %q", iface)
```
Comment (MiguelPires) [inline overlord/ifacestate/apparmorprompting/common/common.go:56]: nitpick: we don't need to (and usually don't) quote `iface` since it comes after the colon IMO. There's other cases of this in this file if you end up changing it

**Review by MiguelPires — APPROVED** [review]

File: overlord/ifacestate/apparmorprompting/common/common.go (hunk 14 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:109]: I like your reasoning, but I would like to leave the Unset values as `""`, since when patching existing rules, only modified fields need to be included (others are omitempty), and by using `""` the default value when unmarshalling from an omitted outcome or lifespan type are the Unset variants.

Comment (MiguelPires) [reply overlord/ifacestate/apparmorprompting/common/common.go:109]: That's a good point. Thanks

File: overlord/ifacestate/apparmorprompting/common/common.go (hunk 13 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:96]: Will fix with future commit to address other review comments.

**Review by olivercalder — COMMENTED** [review]
Thanks for the comments. I'll fix the typo with any other review comments by @zyga @bboozzoo or @pedronis, and perhaps they have preferences regarding the use of `%q` in `ValidateForInterface`.

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 17]
```diff
@@ -0,0 +1,466 @@
...
+	"encoding/base32"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+	"github.com/snapcore/snapd/strutil"
+)
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:32]: this package doesn't depend on overlord/state so it's unclear why it would live here, also the name is a bit too long. Also it seems that some helpers might belong to timeutil or snap/naming. Most of the rest should go to a subpackage of interfaces, like interfaces/prompting perhaps?

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+
+var (
+	ErrInvalidSnapLabel           = errors.New("the given label cannot be converted to snap")
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:35]: snapd tends to call these security tags I think, not label

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+type Constraints struct {
+	PathPattern string   `json:"path-pattern,omitempty"`
+	Permissions []string `json:"permissions,omitempty"`
+}
+
+// ValidateForInterface returns nil if the constraints are valid for the given
+// interface, otherwise returns an error.
+func (c *Constraints) ValidateForInterface(iface string) error {
+	switch iface {
+	case "home", "camera":
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:50]: why camera here already?

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 20]
```diff
@@ -0,0 +1,466 @@
...
+	LifespanForever  LifespanType = "forever"
+	LifespanSession  LifespanType = "session"
+	LifespanSingle   LifespanType = "single"
+	LifespanTimespan LifespanType = "timespan"
+)
+
+// TimestampToTime converts the given timestamp string to a time.Time in Local
+// time. The timestamp string is expected to be of the format time.RFC3339Nano.
+// If it cannot be parsed as such, returns an error.
+func TimestampToTime(timestamp string) (time.Time, error) {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:139]: feels like this belong to timeutil as it doesn't have anything to prompting?  same for a few of the helpers here?

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 21]
```diff
@@ -0,0 +1,466 @@
...
+	id = base32.StdEncoding.EncodeToString(nowBytes)
+	timestamp = now.Format(time.RFC3339Nano)
+	return id, timestamp
+}
+
+// LabelToSnap extracts the snap name from the given AppArmor label.
+//
+// If the label is not of the form 'snap.<snap>.<app>', returns an error, and
+// returns the label as the snap.
+func LabelToSnap(label string) (string, error) {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:181]: this probably belongs to snap/naming as SecurityTag helper?

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 22]
```diff
@@ -0,0 +1,466 @@
...
+		"camera": {
+			"access": notify.AA_MAY_WRITE | notify.AA_MAY_READ | notify.AA_MAY_APPEND,
+		},
+	}
+)
+
+// SelectSingleInterface selects the interface with the highest priority from
+// the given list. If none of the given interfaces are included in
+// interfacePriorities, or if the list is empty, return "other".
+func SelectSingleInterface(interfaces []string) string {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:226]: this probably needs to be introduced in the PR that uses it, I'm not sure in isolation whether it's premature or not

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 23]
```diff
@@ -0,0 +1,466 @@
...
+
+// ValidateLifespanParseDuration checks that the given lifespan is valid and
+// that the given duration is valid for that lifespan.
+//
+// If the lifespan is LifespanTimespan, then duration must be a string parsable
+// by time.ParseDuration(), representing the duration of time for which the rule
+// should be valid. Otherwise, it must be empty. Returns an error if any of the
+// above are invalid, otherwise computes the expiration time of the rule based
+// on the current time and the given duration and returns it.
+func ValidateLifespanParseDuration(lifespan LifespanType, duration string) (string, error) {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:418]: why is this one and the following ones not closer to the corresponding types?

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+func AvailablePermissions(iface string) ([]string, error) {
+	available, exist := interfacePermissionsAvailable[iface]
+	if !exist {
+		return nil, fmt.Errorf("cannot get available permissions: unsupported interface: %q", iface)
+	}
+	return available, nil
+}
+
+// AbstractPermissionsFromAppArmorPermissions returns the list of permissions
+// corresponding to the given AppArmor permissions for the given interface.
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:253]: AppArmor permissions  seems to vague here, it seems to relate to something in notify?  Anyway all these AbstractPermissions function probably should be defined in a separate PR

File: overlord/ifacestate/apparmorprompting/common/common.go [hunk 25]
```diff
@@ -0,0 +1,466 @@
...
+	binary.BigEndian.PutUint64(nowBytes, nowUnix)
+	id := base32.StdEncoding.EncodeToString(nowBytes)
+	return id
+}
+
+// NewIDAndTimestamp returns a new unique ID and corresponding timestamp.
+//
+// The ID is the current unix time in nanoseconds encoded as a string in base32.
+// The timestamp is the same time, encoded as a string in time.RFC3339Nano.
+func NewIDAndTimestamp() (id string, timestamp string) {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/common/common.go:167]: this could either go to timeutil or maybe be introduced when they are used? it's a bit uncler if they need to be exported or not

**Review by pedronis — CHANGES_REQUESTED** [review]
this PR is mixing helpers from too different topics, it needs splitting. As mentioned in my first comment some of the helpers might belong to preexisting packages. Also none of these seem related to/use overlord/state so anything that cannot be put in a preexisting package probably belongs to a subpackage under interfaces

File: overlord/ifacestate/apparmorprompting/common/common.go (hunk 19 above)
Comment (pedronis) [reply overlord/ifacestate/apparmorprompting/common/common.go:50]: we probably we will use tags once camera entire the picture. I think the code here should be simplified and not try to guess too much what we are going to do when more interfaces enter the picture

File: overlord/ifacestate/apparmorprompting/common/common.go (hunk 17 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:32]: Very good point. This also makes much more sense when we want to expose over the API which abstract permissions are available for a given interface.

**Review by olivercalder — COMMENTED** [review]
New plan: split this PR into two, and move to `interfaces/prompting` package

- One PR for all of the helpers which do not relate to abstract permissions, in `interfaces/prompting` package
  - Replace `LabelToSnap` with helper in `snap/info.go`
  - Probably move timestamp functions to `timeutil`
- One PR to add all the abstract permission handling functions to `interfaces/prompting`, on top of previous PR
- Move the path pattern PR to `interfaces/prompting` as well

I'll work on implementing these changes next.

File: overlord/ifacestate/apparmorprompting/common/common.go (hunk 22 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/common/common.go:226]: No, you're right, we don't know anything about how we'll want to go about deciding which interface to use if there are multiple. I'll remove this for now.
